manubot/cite/csl_item.py

```diff
@@ -275,3 +301,75 @@ def standardize_id(self):
 def assert_csl_item_type(x):
     if not isinstance(x, CSL_Item):
         raise TypeError(f"Expected CSL_Item object, got {type(x)}")
+
+
+def date_to_date_parts(iso_date):
+    """
+    Convert a date string or object to a date parts list.
+
+    iso_date: date either as a string (in the form YYYY, YYYY-MM, or YYYY-MM-DD)
+        or as a Python date object (datetime.date or datetime.datetime).
+    """
+    import datetime
+
+    if iso_date is None:
+        return None
+    if isinstance(iso_date, (datetime.date, datetime.datetime)):
+        iso_date = iso_date.isoformat()
+    if not isinstance(iso_date, str):
+        raise ValueError(f"date_to_date_parts: unsupported type for {iso_date}")
+    iso_date = iso_date.strip()
+    re_year = r"(?P<year>[0-9]{4})"
+    re_month = r"(?P<month>1[0-2]|0[1-9])"
+    re_day = r"(?P<day>[0-3][0-9])"
+    patterns = [
+        f"{re_year}-{re_month}-{re_day}",
+        f"{re_year}-{re_month}",
+        f"{re_year}",
+        f".*",  # regex to match anything
+    ]
+    for pattern in patterns:
+        match = re.match(pattern, iso_date)
+        if match:
+            break
+    date_parts = []
+    for part in "year", "month", "day":
+        try:
+            value = match.group(part)
+        except IndexError:
+            break
+        if not value:
+            break
+        date_parts.append(int(value))
+    if date_parts:
+        return date_parts
+
+
+def date_parts_to_string(date_parts, fill: bool = False):
+    """
+    Return a CSL date-parts list as ISO formatted string:
+    ('YYYY', 'YYYY-MM', 'YYYY-MM-DD', or None).
+
+    date_parts: list or tuple like [year, month, day] as integers.
+    fill: if True, set missing months to January
+        and missing days to the first day of the month.
+    """
+    if not date_parts:
+        return None
+    if not isinstance(date_parts, (tuple, list)):
+        raise ValueError(f"date_parts must be a tuple or list")
+    while fill and 1 <= len(date_parts) < 3:
+        date_parts.append(1)
+    widths = 4, 2, 2
+    str_parts = []
+    for i, part in enumerate(date_parts[:3]):
+        width = widths[i]
+        if isinstance(part, int):
+            part = str(part).zfill(width)
```

Comment:
Could this cause unexpected behavior if somehow a year is stored as `YY` (e.g. `98`) and is then zero filled?

Reply:
I am okay with `[99, 1, 1]` becoming `0099-01-01`. Rarely a reference may be to a work of that age. I'd say it's not our problem... users should pass full years. For example, would `10` be 1910 or 2010... I don't think we should make inferences here.

Reply:
I'm okay with this behavior too.  It could be a good test case as a way of documenting that is the expected behavior.